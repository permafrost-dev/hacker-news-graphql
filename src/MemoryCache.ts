export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache {
    public map: CacheMap = {};

    public has(key: string): boolean {
        this.purge();

        return typeof this.map[key] !== 'undefined';
    }

    public get(key: string): any {
        this.purge();

        return this.map[key]?.value ?? null;
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[key] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }
}
