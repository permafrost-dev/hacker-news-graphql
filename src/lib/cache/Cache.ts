/* eslint-disable no-unused-vars */

export abstract class Cache {
    public prefix = '';

    public destroy() {
        //
    }
    public abstract has(key: string): Promise<boolean>;

    public abstract get(key: string): Promise<any | null>;

    public abstract put(key: string, value: any, ttlSeconds: number): void;

    public abstract purge();

    public abstract clear(): boolean;
}
