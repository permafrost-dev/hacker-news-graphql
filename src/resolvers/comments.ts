/* eslint-disable no-unused-vars */

import axios from 'axios';

export function getResolver(cache: any): any {
    const commentsResolver = async (parent, { first }) => {
        const comments: Record<string, any>[] = [];

        if (typeof parent['kids'] === 'undefined') {
            return new Promise(resolve => resolve([]));
        }

        const commentIds = parent.kids.slice(0, 1);

        for (const id of commentIds) {
            if (!cache.has(`comment:${id}`)) {
                // console.log(`getting comment ${id} from URL...`);

                const commentResp = await axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
                const commentData: Record<string, any>[] = commentResp.data;

                cache.put(`comment:${id}`, commentData, 3600);
            }

            comments.push(cache.get(`comment:${id}`));
        }

        return new Promise(resolve => resolve(comments));
    };

    return commentsResolver;
}
