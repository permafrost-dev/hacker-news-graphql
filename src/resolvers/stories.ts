import axios from 'axios';

export function getResolver(topStoryIds, topStories, cache: any): any {
    const stories = async (_, { first, kind }) => {
        topStoryIds.splice(0, topStoryIds.length);
        topStories.splice(0, topStories.length);

        kind = `${kind}`.toLowerCase();

        const ids: number[] = [];
        const topstoryCacheKey = `${kind}storyids:${first}`;

        if (!cache.has(topstoryCacheKey)) {
            // console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/${kind}stories.json?limitToFirst=${first}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(topstoryCacheKey, data, 60);
            ids.push(...data);
        }

        const cacheData = cache.get(topstoryCacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(id => {
                if (cache.has(`${kind}story:${id}`)) {
                    return null;
                }

                // console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map(resp => resp?.data)
            .filter(item => item !== null)
            .forEach(item => {
                cache.put(`${kind}story:${item.id}`, item, 600);
            });

        for (const id of ids) {
            topStories.push(cache.get(`${kind}story:${id}`));
        }

        return new Promise(resolve => resolve(topStories));
    };

    return stories;
}
