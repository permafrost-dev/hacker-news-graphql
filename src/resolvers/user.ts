/* eslint-disable no-unused-vars */

import { MemoryCache } from '../MemoryCache';
import axios from 'axios';

export function getResolver(cache: MemoryCache): any {
    const userResolver = async (_, { id }) => {
        if (cache.has(`user:${id}`)) {
            return cache.get(`user:${id}`);
        }

        const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/user/${id}.json`);

        cache.put(`user:${id}`, resp.data, 3600);

        return cache.get(`user:${id}`);
    };

    return userResolver;
}
