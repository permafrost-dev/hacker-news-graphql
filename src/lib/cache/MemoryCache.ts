import { Cache } from './Cache';

export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache extends Cache {
    public map: CacheMap = {};

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;
    }

    protected makeKey(key: string): string {
        return `${this.prefix}${key}`;
    }

    public async has(key: string): Promise<boolean> {
        this.purge();

        return new Promise(resolve => resolve(typeof this.map[this.makeKey(key)] !== 'undefined'));
    }

    public async get(key: string): Promise<any | null> {
        this.purge();

        const result = this.map[this.makeKey(key)]?.value ?? null;

        return new Promise(resolve => resolve(result));
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[this.makeKey(key)] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }

    public clear() {
        //
    }
}
