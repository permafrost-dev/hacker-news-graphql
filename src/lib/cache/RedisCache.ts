import { Cache } from './Cache';
const Redis = require('ioredis');

export class RedisCache extends Cache {
    public redis: typeof Redis;

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;

        this.redis = new Redis(); //{ keyPrefix: keyPrefix });
    }

    public destroy() {
        this.redis.disconnect();
    }

    public async has(key: string): Promise<boolean> {
        const result = await this.redis.exists(key);

        return new Promise(resolve => resolve(result === 1));
    }

    public async get(key: string): Promise<any | null> {
        const result = await this.redis.get(key);

        return new Promise(resolve => resolve(JSON.parse(result)));
    }

    public put(key: string, value: any, ttlSeconds: number): void {
        this.redis.set(key, JSON.stringify(value), 'EX', ttlSeconds);
    }

    public purge() {
        //
    }

    public clear() {
        //
    }
}
