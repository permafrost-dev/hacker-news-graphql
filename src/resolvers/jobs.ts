import { Cache } from '@/lib/cache/Cache';
import axios from 'axios';
const hash = require('object-hash');

export function getResolver(storyIds, stories, cache: Cache): any {
    const result = async (_, { first, skipText }, { req }) => {
        const kind = 'job';

        storyIds.splice(0, storyIds.length);
        stories.splice(0, stories.length);

        const queryHash = hash(req.body);

        const ids: number[] = [];
        //const cacheKey = `jobstoryids:${first}-${skipText ? 'skipText' : 'dontSkipText'}`;
        const cacheKey = `jobstoryids:${first}:${queryHash}`;

        console.log('cacheKey = ', cacheKey);

        const hasKey = await cache.has(cacheKey);

        if (!hasKey) {
            console.log(`getting ${kind}storyids from URL...`);

            const resp = await axios.get(`${process.env.HACKERNEWS_API_URL}/jobstories.json?limitToFirst=${first + 5}&orderBy="$key"`);
            const data: number[] = resp.data;

            cache.put(cacheKey, data, 600);
            ids.push(...data);
        }

        const cacheData = await cache.get(cacheKey);

        ids.push(...cacheData);
        ids.splice(first);

        const storyDataPromises = ids
            .map(async id => {
                if (await cache.has(`${kind}story:${id}`)) {
                    return cache.get(`${kind}story:${id}`);
                }

                console.log(`getting story ${id} from URL...`);

                return axios.get(`${process.env.HACKERNEWS_API_URL}/item/${id}.json`);
            })
            .filter(item => item !== null);

        (await Promise.all(storyDataPromises))
            .map((resp: any) => resp?.data)
            .filter(item => item !== null)
            .filter(item => typeof item !== 'undefined')
            .forEach(item => {
                try {
                    cache.put(`${kind}story:${item.id}`, item, 3600);
                } catch (err) {
                    console.log(err);
                }
            });
        for (const id of ids) {
            const storyItem = await cache.get(`${kind}story:${id}`);

            if (!skipText || typeof storyItem['text'] === 'undefined' || storyItem['text'] === null) {
                stories.push(storyItem);
            }
        }

        return new Promise(resolve => resolve(stories));
    };

    return result;
}
