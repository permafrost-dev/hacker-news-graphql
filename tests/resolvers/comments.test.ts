/* eslint-disable no-undef */

import { MemoryCache } from '@/lib/cache/MemoryCache';
import { getResolver } from '@/resolvers/comments';

let cache: MemoryCache;
let resolver: any;

process.env.HACKERNEWS_API_URL = 'http://127.0.0.1:3031';

beforeEach(() => {
    cache = new MemoryCache();
    resolver = getResolver(cache);
});

it('it retrieves the comment data and returns it', async () => {
    const data = await resolver({ kids: [27397733] }, { first: 1 });

    expect(data).toMatchSnapshot();
});

it('it retrieves the comment data and caches it', async () => {
    await resolver({ kids: [27397733] }, { first: 1 });

    expect(cache.has('comment:27397733')).toBeTruthy();
});
