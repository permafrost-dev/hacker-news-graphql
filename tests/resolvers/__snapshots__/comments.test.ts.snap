// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it retrieves the comment data and returns it 1`] = `
Array [
  Object {
    "by": "mettamage",
    "id": 27397733,
    "kids": Array [
      27397940,
      27397843,
      27398125,
      27398799,
    ],
    "parent": 27397208,
    "text": "You might want to add (obfuscated) to the title. The actual source code, while small, is much larger [1]. (edit: at 589 lines it&#x27;s still impressive!)<p>For people who like this kind of thing and&#x2F;or for people who want to be able to understand what is going on at all, you might like the course NAND2Tetris [2]. It&#x27;s a good prerequisite to understanding this blog post if computer systems isn&#x27;t a topic that one has explored before.<p>[1] <a href=\\"https:&#x2F;&#x2F;gist.github.com&#x2F;binji&#x2F;395669d45e9005950232043ab4378abe\\" rel=\\"nofollow\\">https:&#x2F;&#x2F;gist.github.com&#x2F;binji&#x2F;395669d45e9005950232043ab4378a...</a> -- the author notes this in the article.<p>[2] <a href=\\"https:&#x2F;&#x2F;www.nand2tetris.org&#x2F;\\" rel=\\"nofollow\\">https:&#x2F;&#x2F;www.nand2tetris.org&#x2F;</a>",
    "time": 1622833637,
    "type": "comment",
  },
]
`;
